src/hardware/mouse/mouse_queue.cpp

```diff
@@ -0,0 +1,317 @@
+		LOG_INFO("(queue) %04d: " fmt, DEBUG_GetDiffTicks(), __VA_ARGS__);
+
+static uint32_t DEBUG_GetDiffTicks()
+{
+	static uint32_t previous_ticks = 0;
+	uint32_t diff_ticks            = 0;
+
+	if (previous_ticks)
+		diff_ticks = PIC_Ticks - previous_ticks;
+
+	previous_ticks = PIC_Ticks;
+	return diff_ticks;
+}
+
+#endif
+
+// ***************************************************************************
+// Mouse event queue implementation
+// ***************************************************************************
+
+void mouse_queue_tick(uint32_t)
+{
+	MouseQueue::GetInstance().Tick();
+}
+
+MouseQueue &MouseQueue::GetInstance()
+{
+	static MouseQueue mouse_queue;
+	return mouse_queue;
+}
+
+MouseQueue::~MouseQueue()
+{
+	PIC_RemoveEvents(mouse_queue_tick);
+}
+
+void MouseQueue::SetRateDOS(const uint16_t rate_hz)
+{
+	// Convert rate in Hz to delay in milliseconds
+	start_delay.dos_ms = MOUSE_GetDelayFromRateHz(rate_hz);
+}
+
+void MouseQueue::SetRatePS2(const uint16_t rate_hz)
+{
+	// Convert rate in Hz to delay in milliseconds
+	start_delay.ps2_ms = MOUSE_GetDelayFromRateHz(rate_hz);
+}
+
+void MouseQueue::AddEvent(MouseEvent &ev)
+{
+	DEBUG_QUEUE("AddEvent:   %s %s",
+	            ev.request_dos         ? "DOS"
+	            : "---" ev.request_ps2 ? "PS2"
+	                                   : "---");
+
+	// Prevent unnecessary processing
+	AggregateDosEvents(ev);
+	if (!ev.request_dos && !ev.request_ps2)
+		return; // event not relevant any more
+
+	bool restart_timer = false;
+	if (ev.request_dos) {
+		if (!HasEventDos() && timer_in_progress && !delay.dos_ms) {
+			DEBUG_QUEUE("AddEvent: restart timer for %s", "DOS");
+			// We do not want the timer to start only then PS/2
+			// event gets processed - for minimum latency it is
+			// better to restart the timer
+			restart_timer = true;
+		}
+
+		if (ev.dos_moved) {
+			// Mouse has moved
+			pending_dos_moved = true;
+		} else if (ev.dos_wheel) {
+			// Wheel has moved
+			pending_dos_wheel = true;
+		} else {
+			// Button press/release
+			pending_dos_button        = true;
+			pending_dos_buttons_state = ev.dos_buttons;
+		}
+	}
+
+	if (ev.request_ps2) {
+		if (!HasEventPS2() && timer_in_progress && !delay.ps2_ms) {
+			DEBUG_QUEUE("AddEvent: restart timer for %s", "PS2");
+			// We do not want the timer to start only when other
+			// event gets processed - for minimum latency it is
+			// better to restart the timer
+			restart_timer = true;
+		}
+
+		// Events for PS/2 interface (or virtual machine compatible
+		// drivers) do not carry any information - they are only
+		// notifications that new data is available
+		pending_ps2 |= ev.request_ps2;
+	}
+
+	if (restart_timer) {
+		timer_in_progress = false;
+		PIC_RemoveEvents(mouse_queue_tick);
+		UpdateDelayCounters();
+		StartTimerIfNeeded();
+	} else if (!timer_in_progress) {
+		DEBUG_QUEUE("ActivateIRQ, in %s", __FUNCTION__);
+		// If no timer in progress, handle the event now
+		PIC_ActivateIRQ(12);
```

Comment:
The magic `12` is used on 148 and 287 -> can you lift it to a local global `constexpr` (Is this the PS/2 mouse IRQ?)

Reply:
OK, added a constant